### ISSUE TITLE
False positive over `@Initializer` static method

### ISSUE THREAD

**Issue by stechio** [uber/NullAway#1625]
NullAway complains about allegedly non-initialized fields despite non-null values are assigned to those fields over all possible paths through a static initialization method marked with `@Initializer`:

```
[WARNING] ...java:[501,48] [NullAway] @NonNull field XXXXX not initialized
```

In particular, in my case I need to deterministically defer the initialization of fields within enum constants because of complex logic shared among them; therefore, I marked with `@Initializer` my static initialization method (`def`):

```java
enum MyEnum {
  CONST1,
  CONST2,
  . . .;

  static {
    /*
      HERE complex state is calculated, so it can be shared among certain enum constants. 
     */
    Object sharedValue1;
    Object sharedValue2;
    Object sharedValue3;
    . . .

    def(CONST1, v1, sharedValue2, v3); 
    def(CONST2, sharedValue1, sharedValue2, v3b);
    . . .
  }

  @Initializer
  private static void def(MyEnum e, Object myField1, Object myField2, Object myField3) {
    e.myField1 = myField1;
    e.myField2 = myField2;
    e.myField3 = myField3;
  }

  private Object myField1; // <-- [NullAway] @NonNull field myField1 not initialized
  private Object myField2; // <-- [NullAway] @NonNull field myField2 not initialized
  private Object myField3; // <-- [NullAway] @NonNull field myField3 not initialized
}
```

It seems that NullAway ignores static methods as initializers: can someone confirm whether such methods are supported?

**Comment by Shankar-v27** [comment]
Hi, I would like to work on this issue.

I looked into the behavior and it seems the current @Initializer handling recognizes instance initialization flows but does not account for static initializer helpers that initialize fields through an explicit receiver parameter (for example enum constants).

I will investigate how NullAway tracks initialization state for @Initializer methods and try adding support with a regression test covering static enum initialization.

Could you please assign this to me?

**Comment by msridhar** [comment]
Hi @stechio sorry for my slow response here.  I think your code example is going to be rather challenging to handle in NullAway.  I think the code is safe because the `def` method is called once for each enum value of `MyEnum` in the static initializer block, is that correct?  To really prove this code safe goes beyond supporting `@Initializer` on static methods; we also have to reason that _all_ the correct `def` calls are present in the initializer block.  This would be complicated and rather specialized support for a case that I don't think comes up all that often (though maybe I'm wrong).

Have I understood your example correctly?  If so, would you be ok with just suppressing for this case instead?  You can write `@SuppressWarnings("NullAway.Init")` on each field declaration or on `MyEnum` to only suppress initialization warnings.

**Comment by stechio** [comment]
No problem: I suspected it was a niche case whose complexity wouldn't be deemed worth its maintenance in NullAway, so in the meantime I had already settled on suppression.

Thanks @msridhar for your feedback, and for your dedication to the project!
